File: autocontext/src/autocontext/storage/goal_run_report_store.py
"""File helpers for goal-run reports."""

from __future__ import annotations

from pathlib import Path, PurePosixPath, PureWindowsPath
from typing import Any, Protocol

from autocontext.analytics.goal_run_report import GoalRunReport
from autocontext.util.json_io import read_json, write_json


class DictSerializable(Protocol):
    def to_dict(self) -> dict[str, Any]: ...


def goal_run_report_path(knowledge_root: Path, goal_id: str, goal_run_id: str) -> Path:
    return knowledge_root / "goal_runs" / _path_segment(goal_id, "goal_id") / f"{_path_segment(goal_run_id, 'goal_run_id')}.json"


def _path_segment(value: str, label: str) -> str:
    normalized = value.strip()
    if not normalized:
        raise ValueError(f"{label} is required")
    if "/" in normalized or "\\" in normalized:
        raise ValueError(f"{label} must be a single path segment: {value!r}")
    for path_cls in (PurePosixPath, PureWindowsPath):
        candidate = path_cls(normalized)
        if candidate.is_absolute() or len(candidate.parts) != 1 or candidate.parts[0] in {".", ".."}:
            raise ValueError(f"{label} must be a single path segment: {value!r}")
    return normalized


def write_goal_run_report(knowledge_root: Path, goal_id: str, goal_run_id: str, report: DictSerializable) -> Path:
    path = goal_run_report_path(knowledge_root, goal_id, goal_run_id)
    write_json(path, report.to_dict())
    return path


def read_goal_run_report(knowledge_root: Path, goal_id: str, goal_run_id: str) -> GoalRunReport | None:
    path = goal_run_report_path(knowledge_root, goal_id, goal_run_id)
    return GoalRunReport.from_dict(read_json(path)) if path.exists() else None


__all__ = [
    "goal_run_report_path",
    "read_goal_run_report",
    "write_goal_run_report",
]

File: ts/src/knowledge/goal-run-report-store.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { basename, dirname, join } from "node:path";
import { parseGoalRunReport, type GoalRunReport } from "../analytics/goal-run-report.js";

export function goalRunReportPath(knowledgeRoot: string, goalId: string, goalRunId: string): string {
  return join(
    knowledgeRoot,
    "goal_runs",
    pathSegment(goalId, "goalId"),
    `${pathSegment(goalRunId, "goalRunId")}.json`,
  );
}

function pathSegment(value: string, label: string): string {
  const normalized = value.trim();
  if (!normalized) throw new Error(`${label} is required`);
  if (
    normalized === "." ||
    normalized === ".." ||
    normalized.includes("/") ||
    normalized.includes("\\") ||
    basename(normalized) !== normalized
  ) {
    throw new Error(`${label} must be a single path segment`);
  }
  return normalized;
}

export function writeGoalRunReport(
  knowledgeRoot: string,
  goalId: string,
  goalRunId: string,
  report: GoalRunReport,
): string {
  const path = goalRunReportPath(knowledgeRoot, goalId, goalRunId);
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, JSON.stringify(report, null, 2) + "\n", "utf-8");
  return path;
}

export function readGoalRunReport(knowledgeRoot: string, goalId: string, goalRunId: string): GoalRunReport | null {
  const path = goalRunReportPath(knowledgeRoot, goalId, goalRunId);
  return existsSync(path) ? parseGoalRunReport(JSON.parse(readFileSync(path, "utf-8")) as unknown) : null;
}
